org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceExtension_ServiceAwareTest.java

```diff
@@ -166,53 +231,60 @@ public void matchByFilter() throws Exception {
 
 		schedule(afoo, "foo", "bar");
 		futureAssertThatTest(ServiceWithFilter.class).doesNotThrowAnyException();
-		ServiceAware<Foo> serviceAware = TestBase.lastServiceAware.get();
-		SoftAssertions softly = new SoftAssertions();
-
-		softly.assertThat(serviceAware.getService())
-			.as("getService %s", serviceAware)
-			.isNotNull();
-		softly.assertThat(serviceAware.getServiceReference())
-			.as("getServiceReference %s", serviceAware)
-			.isNotNull();
-		softly.assertThat(serviceAware.getServiceReferences())
-			.as("getServiceReferences %s", serviceAware)
-			.isNotNull();
-		softly.assertThat(serviceAware.getServices())
-			.as("getServices %s", serviceAware)
-			.contains(afoo);
-		softly.assertThat(serviceAware.getTimeout())
-			.as("getTimeout %s", serviceAware)
-			.isEqualTo(DEFAULT_TIMEOUT);
-		softly.assertThat(serviceAware.getTracked())
-			.as("waitForService %s", serviceAware)
-			.isNotEmpty();
-		softly.assertThat(serviceAware.getTrackingCount())
-			.as("getTrackingCount %s", serviceAware)
-			.isGreaterThanOrEqualTo(1);
-		softly.assertThat(serviceAware.getCardinality())
-			.as("getCardinality %s", serviceAware)
-			.isEqualTo(1);
-		softly.assertThat(serviceAware.size())
-			.as("size %s", serviceAware)
-			.isEqualTo(1);
-		softly.assertThat(serviceAware.isEmpty())
-			.as("isEmpty %s", serviceAware)
-			.isFalse();
-		softly.assertThat(serviceAware.waitForService(20))
-			.as("waitForService %s", serviceAware)
-			.isNotNull();
+		SoftAssertions softly = TestBase.lastSoftAssertions.get();
+		Foo service = TestBase.lastService.get();
+
+		softly.assertThat(service)
+			.as("getService %s", service)
+			.isSameAs(afoo);
 
 		softly.assertAll();
 	}
 
 	static class ServiceWithCardinality2 extends TestBase {
 		@InjectService(cardinality = 2)
-		ServiceAware<Foo> foo;
+		ServiceAware<Foo> serviceAware;
+
+		@Override
+		Foo getService() {
+			return serviceAware.getService();
+		}
+
+		@Override
+		List<Foo> getServices() {
+			return serviceAware.getServices();
+		}
 
 		@Override
-		ServiceAware<Foo> getServiceAware() {
-			return foo;
+		@Test
+		void test() throws Exception {
+			softly.assertThat(serviceAware.getServiceReference())
+				.as("getServiceReference %s", serviceAware)
+				.isNotNull();
+			softly.assertThat(serviceAware.getServiceReferences())
```

Comment:
I'm curious: What was the reasoning behind moving these here?

Reply:
The tests ran a test case and then, after the test case completed, examined the ServiceAware provided to the test case. But that violates the lifecycle of the ServiceAware. It should not be valid after the test case completes since we should be clean up resources such as closing the underlying ServiceTracker. This PR fixes the problems with not properly closing the ServiceConfiguration(ServiceAware) and thus the tests had to be modified accordingly.